File: functions/shared/string_utils.py
# Copyright 2025 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================

import re

_ARXIV_ID_PREFIX = "http://arxiv.org/abs/"
_ID_VERSION_DELIMETER = "v"


def get_versioned_id(arxiv_id: str, version: str):
    if not version:
        raise ValueError("Empty version")
    if not arxiv_id:
        raise ValueError("Empty id")

    return f"{arxiv_id}{_ID_VERSION_DELIMETER}{version}"


def get_id_and_version(versioned_id):
    if not "v" in versioned_id:
        raise ValueError("Invalid arxiv link")

    split = versioned_id.split(_ID_VERSION_DELIMETER)
    if len(split) != 2:
        raise ValueError("Could not parse single id and version")

    [arxiv_id, version] = split

    if not arxiv_id.strip() or not version.strip():
        raise ValueError("Parsed invalid id")

    return arxiv_id, version


def get_arxiv_versioned_id(metadata_id_link: str):
    if not metadata_id_link.startswith(_ARXIV_ID_PREFIX):
        raise ValueError("Invalid arxiv link")

    return metadata_id_link[len(_ARXIV_ID_PREFIX) :]


def extract_json_from_decorator(text: str) -> str:
    """
    Extracts JSON content from a string formatted as 'json```{json_content}```'.

    Args:
        text: The input string potentially containing the JSON content.

    Returns:
        The parsed json_content string if found, otherwise returns the original text.
    """
    # (.*?) This capturing group contains: (1) `.` to matches any character (usually except
    # for line terminators, except by passing re.DOTALL we include newlines as well)
    # and (2) *? matches the previous dot token between zero and unlimited times, as
    # few times as possible (lazy).
    pattern = r"```json(.*?)```"

    match = re.search(pattern, text, re.DOTALL)

    if match:
        return match.group(1).strip()
    return text

File: frontend/src/shared/string_utils.ts
/**
 * @license
 * Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// Utility functions for processing strings

// Matches "arxiv.org/{some number of letters...}/####.#####
// [a-zA-Z]+: greedily matches letters bewteen 1 and unlimited times
// (\d{4}\.\d{5}): capturing group for id in format '####.#####' where # is a digit
// (?!\d): negative lookahead that checks the next character is NOT a digit
const ARXIV_CAPTURE_REGEX = /arxiv\.org\/[a-zA-Z]+\/(\d{4}\.\d{5})(?!\d)/;

/**
 * Extracts an arXiv ID from a URL or string.
 *
 * Handles various formats like:
 * - https://arxiv.org/pdf/1511.02799
 * - https://arxiv.org/abs/1511.02799
 * - https://arxiv.org/pdf/1511.02799.pdf
 * - https://arxiv.org/pdf/1511.02799v3
 * - arxiv.org/pdf/1511.02799
 *
 * @param url The string to parse.
 * @returns The arXiv ID (e.g., "1511.02799") or null if not found.
 */
export function extractArxivId(url: string): string | null {
  const match = ARXIV_CAPTURE_REGEX.exec(url);
  if (!match) return null;

  const id = match[1]; // First capturing group
  if (!id) return null;

  return id;
}

/**
 * Parses a string of the form "key: value" into a key and value.
 */
export function parseColonKeyValue(originalString: string): {
  key: string;
  value: string;
} {
  const split = originalString.split(":");
  const key = split[0].trim();
  const value = split.slice(1).join(":").trim();
  return { key, value };
}
